vscode-dotty/src/tracer.ts

```diff
@@ -0,0 +1,281 @@
+import * as request from 'request';
+
+import { TracingConsentCache } from './tracing-consent';
+
+export interface Ctx {
+  readonly extensionContext: vscode.ExtensionContext,
+  readonly extensionConfig: vscode.WorkspaceConfiguration,
+  readonly extensionOut: vscode.OutputChannel
+}
+
+export class Tracer {
+    private readonly ctx: Ctx
+
+    private projectId: string
+    private machineId: string
+    private sessionId: string
+
+    private tracingConsent: TracingConsentCache
+
+    private remoteTracingUrl?: string
+
+    constructor(ctx: Ctx) {
+        this.ctx = ctx;
+
+        this.tracingConsent = new TracingConsentCache(ctx.extensionContext.workspaceState);
+
+        this.machineId = (() => {
+            const machineIdKey = 'tracing.machineId';
+            function persisted(value: string): string {
+                ctx.extensionConfig.update(machineIdKey, value, vscode.ConfigurationTarget.Global)
+                return value
+            }
+
+            const machineId = ctx.extensionConfig.get<string | null>(machineIdKey)
+            if (machineId != null) return machineId;
+
+            // vscode.env.machineId is a dummy value if telemetry is off - cannot be used
+            const vscodeMachineId = vscode.workspace.getConfiguration().get<string>('telemetry.machineId')
+            if (vscodeMachineId !== undefined) return persisted(vscodeMachineId)
+
+            function uuidv4() {
+                // https://stackoverflow.com/a/2117523
+                return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function (c) {
+                    var r = Math.random() * 16 | 0, v = c == 'x' ? r : (r & 0x3 | 0x8);
+                    return v.toString(16);
+                });
+            }
+
+            return persisted(uuidv4())
+        })();
+
+        this.projectId = vscode.workspace.name !== undefined ? vscode.workspace.name : 'no-project';
+        this.sessionId = new Date().toISOString();
+    }
+
+    initializeAsyncWorkspaceDump() {
+        const remoteWorkspaceDumpUrl = this.ctx.extensionConfig.get<string>('remoteWorkspaceDumpUrl');
+        if (remoteWorkspaceDumpUrl === undefined) return;
+
+        try {
+            this.asyncUploadWorkspaceDump(remoteWorkspaceDumpUrl);
+        } catch (err) {
+            this.logError('error during workspace dump', safeError(err));
+        }
+    }
+
+    createLspOutputChannel(): vscode.OutputChannel | undefined {
+        const remoteTracingUrl = this.ctx.extensionConfig.get<string>('remoteTracingUrl');
+        if (!remoteTracingUrl) return undefined;
+
+        if (this.tracingConsent.get() === 'no-answer') {
+            vscode.window.showInformationMessage(
+                'Do you want to help EPFL develop this plugin by uploading your usage data? ' +
+                'PLEASE BE AWARE that this will upload all of your keystrokes and all of your code, ' +
+                'among other things.',
+                'yes', 'no'
```

Comment:
I think that just saying "among other things" is scary. We should either exhaustively list what's being uploaded here or include a link to a page that exhaustively lists what's being uploaded.

Also, I would replace "all of your keystrokes" by "every action performed in Dotty IDE". Keystrokes outside of the IDE are not recorded.